scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java

```diff
@@ -18,167 +18,149 @@
-	 */
-	private double runGraphicsMagickOperation(final IMOperation gmOperation) {
 		try {
-			gmConsole.run(gmOperation);
-			return getRmaeValueFromOutput();
-		} catch (final Exception e) {
-			LOGGER.warn("Graphics Magick operation failed. Default screenshot changerate '"
-					+ SCREENSHOT_DEFAULT_CHANGE_RATE + "' gets returned.");
-			LOGGER.warn("gmoperation:" + gmOperation.toString());
-			LOGGER.warn("EXCEPTION: ", e);
-			if (gmConsoleErrorConsumer.getOutput().size() > 0) {
-				final String errorMessage = gmConsoleErrorConsumer.getOutput().get(0);
-				LOGGER.warn(errorMessage);
-			}
-			return SCREENSHOT_DEFAULT_CHANGE_RATE;
+			int diffColor = new Color(255, 0, 0, 200).getRGB();
 
-		} finally {
-			gmConsoleOutputConsumer.clear();
-		}
-	}
+			BufferedImage oldImage = ImageIO.read(baseScreenshot);
+			BufferedImage newImage = ImageIO.read(comparisonScreenshot);
 
-	/**
-	 * Reads the result from the GraphicsMagick console and parses the difference value.
-	 */
-	double getRmaeValueFromOutput() {
-		final ArrayList<String> gmConsoleOutput = gmConsoleOutputConsumer.getOutput();
-		String total = null;
-		for (final String line : gmConsoleOutput) {
-			if (line.contains("Total")) {
-				total = line;
+			// create images with the maximum required size to ensure that both images fit
+			int unifiedWidth = Math.max(oldImage.getWidth(), newImage.getWidth());
+			int unifiedHeight = Math.max(oldImage.getHeight(), newImage.getHeight());
+
+			// convert images to ARGB format to ensure that the data buffers store one pixel as ARGB
+			BufferedImage oldImageArgb = new BufferedImage(unifiedWidth, unifiedHeight, BufferedImage.TYPE_INT_ARGB);
+			BufferedImage newImageArgb = new BufferedImage(unifiedWidth, unifiedHeight, BufferedImage.TYPE_INT_ARGB);
+
+			// for simplicity paint both images on the upper left corner of the compare image since searching the sub-image is too expensive
+			// areas not covered by images are transparent which will lead to a maximum difference on all 3 measured color channels
+			oldImageArgb.getGraphics().drawImage(oldImage, 0, 0, null);
+			newImageArgb.getGraphics().drawImage(newImage, 0, 0, null);
+
+			BufferedImage diffImage = new BufferedImage(Math.max(oldImageArgb.getWidth(), newImageArgb.getWidth()), Math.max(oldImageArgb.getHeight(), oldImageArgb.getHeight()), BufferedImage.TYPE_INT_ARGB);
+
+			boolean diffAvailable = false;
+			double absoluteErrorTotal = 0;
+
+			// improve performance by accessing data buffers directly since methods to access pixels through BufferedImage are passed through the ColorModel which slows things down
+			DataBuffer oldImageBuffer = oldImageArgb.getRaster().getDataBuffer();
+			DataBuffer newImageBuffer = newImageArgb.getRaster().getDataBuffer();
+			DataBuffer diffImageBuffer = diffImage.getRaster().getDataBuffer();
+
+			int totalPixels = oldImageBuffer.getSize();
+
+			for (int n = 0; n < oldImageBuffer.getSize(); n++) {
```

Comment:
Maybe could use Java 8 stream api. Doing this work with parallelStream could improve the performance.